### PR TITLE
[RSDK-3647]Renaming Radius Clustering to Obstacles Pointcloud

### PR DESCRIPTION
**Pull request by vpandiarajan20**
This commit does NOT optimize the Radius Clustering algorithm, it simply renames it, adds some parameters, and adds a benchmarking test.

Manually tested using transform and file camera, several configs, several pc files. Unable to perform comprehensive testing with realsense due to hardware issues. 

Also used through python SDK, performed properly.

### CODE REVIEW

**Review by bhaney — CHANGES_REQUESTED** [review]
This looks good, but just add a comment saying what manual testing you performed, and if you can still use this as expected from app.viam.com. If it all works, then just re-request my review and I will approve

**Comment by vpandiarajan20** [comment]
Benchmarking for point cloud clustering
densePointCloud
   After Every Thread: 64589 points 2892968208 ns/op	180655960 B/op	 1027925 allocs/op
   Lock as you go: 64589 points 2932391958 ns/op	176376304 B/op	 1019461 allocs/op
   KD Tree: 64589 points	     1742016166 ns/op	305236656 B/op	 1803121 allocs/op
pointcloud: 54074 points 
   After Every Thread: 1664290584 ns/op	143709120 B/op	  950791 allocs/op
   Lock as you go: 
   KD Tree: 1018528958 ns/op	169480384 B/op	 1499867 allocs/op
pointcloud_data: 14789 points
   After Every Thread: 2422446416 ns/op	231929808 B/op	 1701029 allocs/op
   Lock as you go: 
   KD Tree: 2450854833 ns/op	246082888 B/op	 1939963 allocs/op
pc005522: 47208 points
   After Every Thread: 2245281125 ns/op	211364416 B/op	 1005008 allocs/op
   Lock as you go: 
   KD Tree: 1154025917 ns/op	146835080 B/op	 1400751 allocs/op

File: vision/segmentation/radius_clustering.go
Comment (tessavitabile) [inline vision/segmentation/radius_clustering.go:22]: We should send out a scope update for any config changes.

**Review by tessavitabile — COMMENTED** [review]
Left a comment, but no need to wait for my sign-off. Also, I was curious if you tested with a lidar, since that's how we're expecting this model to be used.

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:216]: I think this function may need to be refactored - finding the planes, and finding the ground plane, are two distinct functions of the the algorithm. Sometimes you'll want to find all planes, and sometimes you'll just want to find the ground plane.

**Review by bhaney — CHANGES_REQUESTED** [review]
I think I would like a re-factor of the plane finding algorithm.

Finding the ground plane requires extra parameters, like the normal vector which represents an ideal ground plane, and the angleThreshold of the matching plane. So I think what would be great is if you could refactor the functions such that:

you split PointCloudPlaneSegmentation into two separate functions- one which does what already happen, just finds all planes, but another that will become the ground plane finding function. If a user is only trying to find the ground plane, then you can actually stop the plane finding algorithm a bit earlier, since you only need to return one plane.

So because there will be code that both functions use, you can pull out the common code into helper functions, but make the GroundPlaneSegmentation its own function.

File: vision/segmentation/radius_clustering.go
Comment (bhaney) [reply vision/segmentation/radius_clustering.go:22]: Sent the email

Comment (tessavitabile) [reply vision/segmentation/radius_clustering.go:22]: thank you!

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:116]: I don't think you need this. I suspect it may extend the time you are searching by quite a bit. Even though many of the planes you match will not fall within the ground plane limits, it's not necessary to always have N many "viable" candidates. May of them will end up being repeats.

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:168]: I don't think you need pointers to slices `*[][4]float64` and `*[]r3.Vector`. something like `[]r3.Vector` is already a pointer to an array -- so you don't have to worry about copying the whole slice of r3.Vector over. Just use `[][4]float64` and `[]r3.Vector`

Comment (bhaney) [reply vision/segmentation/plane_segmentation.go:168]: Same with `*pc.PointCloud`, PointCloud is already an interface, which means it is already a pointer. doing `*pc.PointCloud` is a pointer to a pointer.

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:254]: There should only be 1 ground plane, right? So just call it `FindGroundPlane` and have it return just a pc.Plane rather than []pc.Plane. Like, I could see that there are 4 planes in a scene, two walls, the ground, and table. The ground and the table are both "ground planes" because they match the flat normal of the input vector. But the table should be an obstacle, and shouldn't be counted as a ground plane

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:346]: I would say, you can stop here after you find the first ground plane. I believe the RANSAC algo always finds the biggest plane first, so once you get the biggest one, just return that pc.Plane (and not a slice of pc.Plane) and that's it.

File: vision/segmentation/plane_segmentation.go
Comment (bhaney) [inline vision/segmentation/plane_segmentation.go:344]: you can return nil plane if there is not a ground plane in the image. It will be up to the function getting the result to check and figure out what to do if there are no ground planes.

**Review by bhaney — CHANGES_REQUESTED** [review]
Nice work! I think there are some places you can speed things up, and also change the signature of the new GroundPlane function.

File: vision/segmentation/plane_segmentation.go
Comment (vpandiarajan20) [reply vision/segmentation/plane_segmentation.go:168]: Got it, didn't know that applied for interfaces actually!

File: vision/segmentation/plane_segmentation.go
Comment (vpandiarajan20) [reply vision/segmentation/plane_segmentation.go:254]: Makes sense, I think the idea was if your maxDistFromPlane is small, you would still want to catch stragglers/discrepancies, but hopefully they would be considered noise and filtered later anways.

File: vision/segmentation/plane_segmentation.go
Comment (vpandiarajan20) [reply vision/segmentation/plane_segmentation.go:116]: Agreed, but since the array of equations is already made, we will probably be looping over all entries in the findBestEq function. I assume a lot of the array would then be blank, so I'm going to skip the equation 0x + 0y + 0z = 0.

**Review by bhaney — APPROVED** [review]
LGTM! Congrats, thanks for testing this out and making the refactors

**Pull request merged**
